src/_P044_P1WifiGateway.ino

```diff
@@ -5,127 +5,223 @@
+//    P1 wifi gateway shield (see http://www.esp8266thingies.nl for print design and kits)
+//    See also http://domoticx.com/p1-poort-slimme-meter-hardware/
 //#######################################################################################################
 
 #include "_Plugin_Helper.h"
 
 #define PLUGIN_044
 #define PLUGIN_ID_044         44
 #define PLUGIN_NAME_044       "Communication - P1 Wifi Gateway"
-#define PLUGIN_VALUENAME1_044 "P1WifiGateway"
+
+#ifndef PLUGIN_044_DEBUG
+  #define PLUGIN_044_DEBUG                 false // extra logging in serial out
+#endif
 
 #define P044_STATUS_LED                    12
-#define P044_BUFFER_SIZE                   1024
+#define P044_CHECKSUM_LENGTH               4
+#define P044_DATAGRAM_START_CHAR           '/'
+#define P044_DATAGRAM_END_CHAR             '!'
+#define P044_DATAGRAM_MAX_SIZE             1024
 #define P044_NETBUF_SIZE                   128
-#define P044_DISABLED                      0
-#define P044_WAITING                       1
-#define P044_READING                       2
-#define P044_CHECKSUM                      3
-#define P044_DONE                          4
 
+#define P044_WIFI_SERVER_PORT     ExtraTaskSettings.TaskDevicePluginConfigLong[0]
+#define P044_BAUDRATE             ExtraTaskSettings.TaskDevicePluginConfigLong[1]
+#define P044_RX_WAIT              PCONFIG(0)
+#define P044_SERIAL_CONFIG        PCONFIG(1)
+#define P044_RESET_TARGET_PIN     CONFIG_PIN1
+ 
 
+struct P044_Task : public PluginTaskData_base {
 
-struct P044_data_struct : public PluginTaskData_base {
+  enum ParserState : byte {
+    P044_WAITING,
+    P044_READING,
+    P044_CHECKSUM
+  };
 
-  P044_data_struct() {
+  P044_Task() {
     clearBuffer();
   }
 
-  ~P044_data_struct() {
+  ~P044_Task() {
     stopServer();
   }
 
+  inline static bool serverActive(WiFiServer *server) {
+#if defined(ESP8266)
+    return nullptr != server && server->status() != CLOSED;
+#elif defined(ESP32)
+    return nullptr != server && *server;
+#endif
+  }
+
   void startServer(unsigned int portnumber) {
+    if (gatewayPort == portnumber && serverActive(P1GatewayServer)) {
+      // server is already listening on this port
+      return;
+    }
     stopServer();
+    gatewayPort = portnumber;
     P1GatewayServer = new WiFiServer(portnumber);
-    if (nullptr != P1GatewayServer) {
+    if (nullptr != P1GatewayServer && WiFi.isConnected()) {
       P1GatewayServer->begin();
-      addLog(LOG_LEVEL_DEBUG, String(F("P1   : WiFi server started at port ")) + portnumber);
+      if(serverActive(P1GatewayServer)) {
+        addLog(LOG_LEVEL_INFO, String(F("P1   : WiFi server started at port ")) + portnumber);
+      } else {
+        addLog(LOG_LEVEL_ERROR, String(F("P1   : WiFi server start failed at port ")) +
+            portnumber + String(F(", retrying...")));
+      }
+    }
+  }
+
+  void checkServer() {
+    if (!serverActive(P1GatewayServer) && WiFi.isConnected()) {
+      P1GatewayServer->close();
```

Comment:
`serverActive` returns also false if `P1GatewayServer == nullptr`

Reply:
But then the next line will crash the ESP as you try to dereference a null-pointer.

Reply:
Yep, if you consider `P044_Task` as just a bunch of data and methods.
A pointer to an uninitialized task object can only be obtained by `init` and is used only at the `init` state. The `init` state is only successful if the task has been initialized.
In all other cases the task pointer is obtained by method `get` which checks if all pointers are unequal to 0.
Both 'startServer' (of the Gateway) and `serialBegin` (of the Parser) are the means to initialize both parts of the task and are the only ones (apart from internal task usage) that are allowed to be called in an initialized object state.
But if you want I can add additional sanity checks...

Reply:
Both are OK.
Either refactoring the code to be part of the struct that on itself is initialized or not, or have a check for null pointer.

The 'problem' with the current code is that it may work correct as it is used now.
But having a separate function that can be called is inviting to be called from other parts of the code.
So a function must not expect external pre-conditions to exist when it should be called or not.
If there are specific conditions on when to call the function, the function itself should check on it, or you must hide it as a private function in an object. As private function of an object, it is up to the object to check.

Reply:
Sanity check added

Reply:
Can you change `WiFi.isConnected()` to `WiFiConnected()` ?

Reply:
Ok, was not aware of that one.